proxy/coco/src/request/waiting_room.rs

```diff
@@ -0,0 +1,700 @@
+                SomeRequest::Cloning(request) => Some(request),
+                _ => None,
+            },
+            |previous| Ok(previous.failed(peer, timestamp)),
+            urn,
+        )
+    }
+
+    /// Tell the `WaitingRoom` that we successfully cloned the `found_repo` for the given `urn`.
+    ///
+    /// If the underlying `Request` was in the `Cloning` state then it will transition to the
+    /// `Cloned` state.
+    ///
+    /// # Errors
+    ///
+    ///   * If the `urn` was not in the `WaitingRoom`.
+    ///   * If the underlying `Request` was not in the expected state.
+    pub fn cloned(&mut self, urn: &RadUrn, found_repo: RadUrn, timestamp: T) -> Result<(), Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Cloning(request) => Some(request),
+                _ => None,
+            },
+            |previous| Ok(previous.cloned(found_repo, timestamp)),
+            urn,
+        )
+    }
+
+    /// Tell the `WaitingRoom` that we are cancelling the request for the given `urn`.
+    ///
+    /// If the underlying `Request` was in the `{IsCreated, IsRequested, Found, Cloning,
+    /// IsCanceled}` state then it will transition to the `IsCanceled` state.
+    ///
+    /// # Errors
+    ///
+    ///   * If the `urn` was not in the `WaitingRoom`.
+    ///   * If the underlying `Request` was not in the expected state.
+    pub fn canceled(&mut self, urn: &RadUrn, timestamp: T) -> Result<(), Error>
+    where
+        T: Clone,
+    {
+        self.transition(|request| request.cancel(timestamp).right(), Ok, urn)
+    }
+
+    /// Return the list of all `RadUrn`s in the `WaitingRoom`.
+    pub fn list(&self) -> impl Iterator<Item = &RadUrn> {
+        self.requests.keys()
+    }
+
+    /// Filter the requests in the waiting room based on the passed in `matcher`.
+    fn filter<R: Rng, S>(
+        &self,
+        mut matcher: impl FnMut(&SomeRequest<T>) -> Option<&Request<S, T>>,
+        strategy: Strategy<R>,
+    ) -> Option<&Request<S, T>>
+    where
+        T: Clone + Ord,
+    {
+        strategy.next(
+            self.requests
+                .iter()
+                .filter_map(|(_, request)| matcher(request)),
+            |request| request.timestamp.clone(),
+        )
+    }
+
+    /// Return a `Request` that is in the `IsCreated` state, if any, based off of the supplied
+    /// `strategy`.
+    pub fn next<R: Rng>(&self, strategy: Strategy<R>) -> Option<&Request<IsCreated, T>>
+    where
+        T: Clone + Ord,
+    {
+        self.filter(
+            |request| match request {
+                SomeRequest::Created(request) => Some(request),
+                _ => None,
+            },
+            strategy,
+        )
+    }
+
+    /// Return a `Request` that is in the `Cloned` state, if any, based off of the supplied
+    /// `strategy`.
+    pub fn ready<R: Rng>(&self, strategy: Strategy<R>) -> Option<&Request<Cloned, T>>
```

Comment:
What's the reason to have `ready` and `next`?

Reply:
Can you elaborate on the question, please :)

Reply:
To expand, probably two streams (in the general sense) are need, one that provides the next candidate to query and another which provides the next candidate to clone.